### PR TITLE
Implement PEP 658: .whl.metadata files

### LINKED ISSUE #3980: Implement PEP 658: .whl.metadata files
To make Simple JSON API more useful(https://github.com/pyodide/pyodide/issues/3979, ), it would be good to deploy the .whl.metadata files per [PEP 658](https://peps.python.org/pep-0658/), so that Python packages manager could more easily identify dependencies. 

This can be done by for instance,
 1. adding a binary  `--metadata-files` option to [`pyodide build-recipes`](https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyodide_build/cli/build_recipes.py#L11) (default to False for backward compatibility)
 2. set to True in the [packages/Makefile](https://github.com/pyodide/pyodide/blob/main/packages/Makefile#L16)
 3. somewhere around [here](https://github.com/pyodide/pyodide/blob/main/pyodide-build/pyodide_build/buildpkg.py#L862) once the wheel is created,  as described in the PEP, extract the METADATA file, and write it in a file named identically to the wheel but with a `.metadata` suffix (see PEP).
 4. Check that such files will be deployed the to CDN with the correct content type by adding another extension check [here](https://github.com/pyodide/pyodide/blob/main/tools/deploy_s3.py#L135) and setting it to the appropriate content type (probably `text/plain` ?)
 5. Ask for this new file extension to be whitelisted in the CDN (I can do it).

@juntyr this could be a good first issue making small changes to different parts of the build system if you are interested.

### PR DESCRIPTION
**Pull request by juntyr**
### Description

To make Simple JSON API more useful (https://github.com/pyodide/micropip/issues/62, https://github.com/pyodide/pyodide/issues/3979), we now deploy the .whl.metadata files per [PEP 658](https://peps.python.org/pep-0658/), so that Python packages manager could more easily identify dependencies.

A `--metadata-files` boolean option is added to the `pyodide build-recipes` command, which defaults to `False` for backward compatibility. When the option is passed alongside the `--install` option and without the `--no_deps` option, the METADATA file of each wheel file `FILENAME.whl` is now installed alongside in `FILENAME.whl.metadata`.

Fixes #3980

### Checklists

- [x] Add a [CHANGELOG](https://github.com/pyodide/pyodide/blob/main/docs/project/changelog.md) entry
- [x] Add / update tests
- [x] Add new / update outdated documentation

### CODE REVIEW

**Comment by juntyr** [comment]
@juntyr TODO: handle the build all case to propagate the metadata file extraction option

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:555]: There is no need to unzip the whole thing. You can use [zipfile](https://docs.python.org/3/library/zipfile.html) to only load the one relevant file. A wheel is  zip file.

File: tools/deploy_s3.py
Comment (rth) [inline tools/deploy_s3.py:140]: Maybe,
```suggestion
            elif file_path.name.endswith('.whl.metadata'):
```
to be a bit more specific.

File: pyodide-build/pyodide_build/buildpkg.py
Comment (rth) [inline pyodide-build/pyodide_build/buildpkg.py:478]: Can probably give it a default value,
```suggestion
    extract_metadata_file: bool = False
```

**Comment by juntyr** [comment]
@rth Thank you for your feedback! I've prototyped two possible implementations.

The first one, based on your suggestions can be seen at https://github.com/pyodide/pyodide/compare/main...juntyr:pyodide:54c482b849846ff01e323824c4af84e2116916ba (note though that it's only 95% complete). It pushes the metadata extraction down to building the package. However, this requires quite invasive code changes. Furthermore, the interaction with installing is a bit weird. What if the package was built with extraction but is installed without metadata? What if it's the other way around?

At the moment, I prefer the second option, which you can see in the current diff. Metadata extraction is now pushed to the installation stage, which means the diff is very clean.

**Review by rth — COMMENTED** [review]
Thanks @juntyr ! No the current version is fine, that was just an approximate suggestion as to where to put it in the code.

Overall LGTM assuming tests pass, but since I was the one who proposed this feature it would be good to have a final review/apporval done by another maintainer.

File: pyodide-build/pyodide_build/cli/build_recipes.py
Comment (juntyr) [reply pyodide-build/pyodide_build/cli/build_recipes.py:102]: In that case, no installation happens, so the extraction is not run. We could still run it and also extract the metadata into the per-package dist-files, but I again prefer not infecting the building code as deeply as option 1 requires.

File: pyodide-build/pyodide_build/cli/build_recipes.py
Comment (ryanking13) [inline pyodide-build/pyodide_build/cli/build_recipes.py:134]: I think we should only show this warning when the metadata-files option is set.

File: pyodide-build/pyodide_build/buildall.py
Comment (ryanking13) [inline pyodide-build/pyodide_build/buildall.py:726]: Could you move this (extracting metadata file from a wheel) into a separate function, and add some tests for the function? I think you can put it in `common.py`.

**Review by ryanking13 — CHANGES_REQUESTED** [review]
Thanks @juntyr for working on this!

File: pyodide-build/pyodide_build/buildall.py
Comment (juntyr) [reply pyodide-build/pyodide_build/buildall.py:729]: FYI, your proposed approach could be implemented as follows:
```python
metadata_infos = [
    info for info in wheel.infolist() if Path(info).name == "METADATA" and len(Path(info).parts) == 2
]

if len(metadata_infos) > 1:
    raise Exception(
        f"Unexpected number of METADATA files when building {pkg.name}: {len(metadata_infos)}"
    )
elif len(metadata_infos) == 1:
    metadata_info = metadata_infos[0]
    metadata_path = metadata_info.filename
    metadata_info.filename = f"{dist_artifact_path.name}.metadata"
    wheel.extract(metadata_path, output_dir)
```

File: pyodide-build/pyodide_build/buildall.py
Comment (juntyr) [reply pyodide-build/pyodide_build/buildall.py:727]: Thank you for the suggestions - I've extracted a helper function and added two tests for it.

File: pyodide-build/pyodide_build/buildall.py
Comment (ryanking13) [reply pyodide-build/pyodide_build/buildall.py]: I think it is okay to raise an error and quit if some packages does not follow the PEP. I believe `pip install` will also fail on those cases.

Comment (ryanking13) [reply pyodide-build/pyodide_build/buildall.py]: You can probably adapt how pip finds a .dist_info dir: https://github.com/pypa/pip/blob/ea727e4d6ab598f34f97c50a22350febc1214a97/src/pip/_internal/utils/wheel.py#L38

Comment (juntyr) [reply pyodide-build/pyodide_build/buildall.py]: > Probably it would be useful to check how pip finds a .dist_info dir: https://github.com/pypa/pip/blob/ea727e4d6ab598f34f97c50a22350febc1214a97/src/pip/_internal/utils/wheel.py#L38

I'm not sure I'm happy with my latest changes, as they increase the complexity a bit. However, we would now use the same functionality as pip.

File: pyodide-build/pyodide_build/tests/test_common.py
Comment (ryanking13) [inline pyodide-build/pyodide_build/tests/test_common.py:154]: nitpick: you can use `pytests.raises` here:

```python
with pytest.raises(SomeException):
    extract_wheel_metadata_file(...)
```

File: pyodide-build/pyodide_build/common.py
Comment (ryanking13) [inline pyodide-build/pyodide_build/common.py:383]: pyodide-build depends on `packaging`, so you can import this function from it:

```py
from packaging.utils import canonicalize_name
```

**Review by ryanking13 — APPROVED** [review]
Thanks, @juntyr!

File: pyodide-build/pyodide_build/tests/test_common.py
Comment (juntyr) [reply pyodide-build/pyodide_build/tests/test_common.py:154]: Thanks for the suggestion

File: pyodide-build/pyodide_build/common.py
Comment (juntyr) [reply pyodide-build/pyodide_build/common.py]: Ah, good catch :)

**Comment by juntyr** [comment]
Are the [CLI docs](https://pyodide.org/en/stable/usage/api/pyodide-cli.html) autogenerated? If so, I probably don't need to update any further documentation.

**Comment by juntyr** [comment]
@ryanking13 @rth Thank you so much for your feedback! This PR should now be ready to me squashed and merged.

**Comment by ryanking13** [comment]
Thanks!

> Are the [CLI docs](https://pyodide.org/en/stable/usage/api/pyodide-cli.html) autogenerated? If so, I probably don't need to update any further documentation.

Yes, it is auto-generated.

**Comment by rth** [comment]
Thanks a lot @juntyr and @ryanking13 for the review!

**Pull request merged**
